### ISSUE TITLE
NPE when switching from 0.9.0-M2 to 0.9.0-M3 

### ISSUE THREAD

**Issue by plokhotnyuk** [circe/circe#816]
Got it in my tests for parsing of ADTs:

```
java.lang.NullPointerException was thrown.
java.lang.NullPointerException
	at com.github.plokhotnyuk.jsoniter_scala.macros.CirceEncodersDecoders$anon$lazy$macro$79$1$$anon$18.configuredDecode(CirceEncodersDecoders.scala:15)
	at io.circe.generic.extras.decoding.ConfiguredDecoder$CaseClassConfiguredDecoder.apply(ConfiguredDecoder.scala:33)
	at io.circe.generic.extras.decoding.ReprDecoder.withDiscriminator(ReprDecoder.scala:68)
	at com.github.plokhotnyuk.jsoniter_scala.macros.CirceEncodersDecoders$anon$lazy$macro$95$1$$anon$19.configuredDecode(CirceEncodersDecoders.scala:17)
	at io.circe.generic.extras.decoding.ConfiguredDecoder$AdtConfiguredDecoder.apply(ConfiguredDecoder.scala:61)
	at io.circe.Decoder.decodeJson(Decoder.scala:52)
	at io.circe.Decoder.decodeJson$(Decoder.scala:52)
	at io.circe.generic.decoding.DerivedDecoder.decodeJson(DerivedDecoder.scala:6)
	at io.circe.Parser.finishDecode(Parser.scala:12)
	at io.circe.Parser.finishDecode$(Parser.scala:9)
	at io.circe.parser.package$.finishDecode(package.scala:5)
	at io.circe.Parser.decode(Parser.scala:26)
	at io.circe.Parser.decode$(Parser.scala:25)
	at io.circe.parser.package$.decode(package.scala:5)
	at com.github.plokhotnyuk.jsoniter_scala.macros.JsonCodecMakerBenchmark.readAdtCirce(JsonCodecMakerBenchmark.scala:283)
	at com.github.plokhotnyuk.jsoniter_scala.JsonCodecMakerBenchmarkSpec.$anonfun$new$2(JsonCodecMakerBenchmarkSpec.scala:87)

```
Changes was trivial: https://github.com/plokhotnyuk/jsoniter-scala/compare/5585096bd269...e4542a2bf571

**Comment by travisbrown** [comment]
Thanks—looking now.

**Comment by travisbrown** [comment]
@plokhotnyuk I'm not 100% sure I want to call this a bug. Given this definition:

```scala
object CirceEncodersDecoders {
  val printer: Printer = Printer.noSpaces.copy(dropNullValues = true, reuseWriters = true)
  implicit val config: Configuration = Configuration.default.withDiscriminator("type")
  implicit val aDecoder: Decoder[A] = deriveDecoder[A]
  implicit val bDecoder: Decoder[B] = deriveDecoder[B]
  implicit val cDecoder: Decoder[C] = deriveDecoder[C]
  implicit val adtDecoder: Decoder[AdtBase] = deriveDecoder[AdtBase]
}
```
`cDecoder` depends on `adtDecoder`, which is a recipe for initialization order problems. It happened to work previously because the `Lazy[ReprDecoder[AdtBase]]` wasn't resolved to a value until decoding was being done, but that changed in the third milestone.

I've changed the behavior back so that this works as written (see #818), but I think we'll need to decide whether this use case should be supported in principle.

**Comment by travisbrown** [comment]
I've confirmed that #818 fixes the issue for your tests, and it'll be available in 0.9.0. Sorry for the inconvenience, and thanks again for the report!
